src/VisualStudio/Core/Def/Implementation/MoveToNamespace/VisualStudioMoveToNamespaceOptionsService.cs

```diff
@@ -2,45 +2,78 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable 
+
 using System;
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.MoveToNamespace;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.MoveToNamespace
 {
     [Export(typeof(IMoveToNamespaceOptionsService)), Shared]
     internal class VisualStudioMoveToNamespaceOptionsService : IMoveToNamespaceOptionsService
     {
+        private const int HistorySize = 3;
+
+        public readonly LinkedList<string?> History = new LinkedList<string?>();
+        private readonly Func<MoveToNamespaceDialogViewModel, bool?> _showDialog;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioMoveToNamespaceOptionsService()
         {
+            _showDialog = viewModel => new MoveToNamespaceDialog(viewModel).ShowModal();
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("RoslynDiagnosticsReliability", "RS0034:Exported parts should be marked with 'ImportingConstructorAttribute'", Justification = "Test constructor")]
+        internal VisualStudioMoveToNamespaceOptionsService(Func<MoveToNamespaceDialogViewModel, bool?> showDialog)
+        {
+            _showDialog = showDialog;
         }
 
         public MoveToNamespaceOptionsResult GetChangeNamespaceOptions(
             string defaultNamespace,
             ImmutableArray<string> availableNamespaces,
-            ISyntaxFactsService syntaxFactsService)
+            ISyntaxFacts syntaxFactsService)
         {
             var viewModel = new MoveToNamespaceDialogViewModel(
                 defaultNamespace,
                 availableNamespaces,
-                syntaxFactsService);
+                syntaxFactsService,
+                History.WhereNotNull().ToImmutableArray());
 
-            var dialog = new MoveToNamespaceDialog(viewModel);
-            var result = dialog.ShowModal();
+            var result = _showDialog(viewModel);
 
             if (result == true)
             {
+                OnSelected(viewModel.NamespaceName);
                 return new MoveToNamespaceOptionsResult(viewModel.NamespaceName);
             }
             else
             {
                 return MoveToNamespaceOptionsResult.Cancelled;
             }
         }
+
+        private void OnSelected(string namespaceName)
+        {
+            if (History.Contains(namespaceName))
```

Comment:
could likely remove this for simplicity.

Reply:
Ah I see, Remove returns bool if it actually removed the value. Safe to just call without checking contains first. Thanks :)